Updated value for hbar, G, kB to reflect current standards

CameronKing opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #17264 

#### Brief description of what is fixed or changed
In June 2019, NIST published an update to the recommended values for physical constants in physics and chemistry. This PR updates the numerical values of the underlying physical constants that changed in this latest revision. In addition, derived constants, such as the Josephson constant, were updated to use their underlying definitions. Furthermore, the unit consistency tests were modified to reflect the updates.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* physics.units
  * Updated the values of physical constants (such as Planck's constant) to reflect the current [2018 CODATA values](https://physics.nist.gov/cuu/pdf/wallet_2018.pdf).
   * More physical constant are derived from their fundamental definitions.
<!-- END RELEASE NOTES -->

Commit:
Updated value for hbar to reflect current standard

In June 2019, the recommended value for the reduced Planck's
constant was set to a fixed quantity. Updating this quantity
necessitated updating the value of G, the Newtonian gravitational
constant, and kB, the Boltzmann constant, for internal consistency.
Furthermore, the unit consistency tests were updated. The values
for the tests were based on the 2018 CODATA values.

asmeurer commented:
Looks like a doctest needs to be updated.

Commit:
Updating amu and doc-tests based on updated physical constants

Updating hbar and G caused a knock-on effect where the definition
of amu changed, so that was updated. In addition in the doc-tests,
solving for the period of Venus was updated to reflect the change
in the value of the Newtonian gravitational constant.

CameronKing commented:
Thanks for your patience with a new contributor! The doctests should be working now.

asmeurer commented on sympy/physics/units/definitions.py:
```diff
@@ -358,7 +359,7 @@
 boltzmann = boltzmann_constant = Quantity("boltzmann_constant")
 boltzmann_constant.set_dimension(energy/temperature)
-boltzmann_constant.set_scale_factor(1.38064852e-23*joule/kelvin)
+boltzmann_constant.set_scale_factor(1.380649e-23*joule/kelvin)
```
Is there a reason this one is fewer digits now?

CameronKing replied:
In the latest update to the Boltzmann constant, it was set to be exactly that value. For reference: https://physics.nist.gov/cgi-bin/cuu/Value?k|search_for=boltzmann+constant .

asmeurer replied:
That URL just loads a blank page for me.

CameronKing replied:
Huh, strange. It's also in the [pdf version](https://physics.nist.gov/cuu/pdf/wallet_2018.pdf) that I linked in the issue.

smichr replied:
If it is an exact value then it should be entered as `Rational('1.380649', 10**23)`

oscargus replied:
That is an interesting aspect. Most likely the numbers are given as decimal numbers, so the floating-point representation is not considered. Considering that there are additional uncertain digits, I'm not sure if it makes a huge difference though.

asmeurer replied:
They're not exact per se, since it's only an approximately measured value. Although some of the unit conversions are in fact exact. For example, the number of feet in a meter is exactly 0.3048, and it does use Rational https://github.com/sympy/sympy/blob/df9965b123b8141a4c75151beeb8db7620c774e9/sympy/physics/units/definitions.py#L216

I think we should keep using float here, or Float if the values have more precision than a Python float allows. At any rate, however, this should be a separate issue from this PR, which is just updating some values.

asmeurer commented on sympy/physics/units/definitions.py:
```diff
@@ -368,7 +369,7 @@
 amu = amus = atomic_mass_unit = atomic_mass_constant = Quantity("atomic_mass_constant")
 atomic_mass_constant.set_dimension(mass)
-atomic_mass_constant.set_scale_factor(1.660539040e-24*gram)
+atomic_mass_constant.set_scale_factor(1.660539060e-24*gram)
```
The document you linked has 1.6605390666. Actually it has 1.66053906660050 if you include the digits in parentheses.

CameronKing replied:
I didn't know if there was a policy on including the uncertainty, so I used the certain digits. I could definitely update with the values in parentheses.

asmeurer replied:
I don't know what the policy is either. If none of the other constants include the uncertainty then we should leave it out.

CameronKing replied:
I'll take a look through the constants definitions and see what the standard is.

CameronKing replied:
Looking through, it was hard to find a good point of comparison since some numbers aren't included to full precision (e.g. elementary charge) and a lot of constants need to be updated.

asmeurer commented on sympy/physics/units/definitions.py:
```diff
@@ -323,17 +323,18 @@
 G = gravitational_constant = Quantity("gravitational_constant", abbrev="G")
 gravitational_constant.set_dimension(length**3*mass**-1*time**-2)
-gravitational_constant.set_scale_factor(6.67408e-11*m**3/(kg*s**2))
+gravitational_constant.set_scale_factor(6.67430e-11*m**3/(kg*s**2))
```
This constant also has some extra digits in parentheses. I don't know if those should be included.

CameronKing replied:
My personal preference would be to not include them since they are usually the most likely to change.

oscargus replied:
Look at this:

https://physics.nist.gov/cgi-bin/cuu/Value?k|search_for=abbr_in!

The additonal digits is the uncertainty and it should be be concatenated, instead, here, the value is between ...30-15 and ...30+15. I did the same interpretation earlier.

oscargus replied:
Non-clickable link, but https://physics.nist.gov/cuu/Constants/index.html and click your way through...

CameronKing replied:
Oh, I see... They aren't additional digits, that's the one sigma window for the last two digits. Thanks for pointing that out! They definitely shouldn't be included then.

smichr commented:
For internal consitency, should `G` be expressed in terms of the Boltzmann constant?

bjodah commented:
Great to see that we're getting the latest updates in, thanks! I think Avogadro's constant also needs to be updated? (perhaps more constants, I knew this one was updated, didn't check the full list). Either way, I think we should make sure not to mix values of constants from different CODATA revsions..

Off-topic for this pull-requst, but: I was thinking, that for reproducibility, it would be nice if we (in the future) had versioned constants in submodules e.g. `sympy.physics.constants.codata2018.*`, what do you think?

CameronKing commented:
It looks like a few constants need to be updated and I agree with @bjodah that different CODATA revisions shouldn't be mixed. Since a full update goes outside of the original intent of this PR, should a new issue and PR be made?

oscargus commented:
Will not having updated all constants break* something? If not, I think this is good to go (from my perspective).

* As in other coefficients getting inconsistent values if they are derived compared to the typed value. I haven't look at this code for a while, but I guess we should try to type a minimal number of constants and derive all other.

CameronKing commented:
I don't know about all constants, but at a minimum, the value for the elementary constant needs to be updated. With the current value of the elementary charge, the derived value of hbar (converted to eV*s from the value in this PR) has an error of 1e-6. For reference, this error is on the same order as updating hbar to the current CODATA revision.

Commit:
Updated electronvolt scale factor

The value of the elementary charge was set to a fixed value as in
the 2018 revision of the CODATA accepted values. This sets the
conversion factor to the full precision of that value.

asmeurer commented:
The atomic mass issue still needs to be addressed https://github.com/sympy/sympy/pull/17273#discussion_r307861804 (not including the uncertainty is fine, but the value you have here is different from the value in the PDF in the last digit, sorry if that wasn't clear in my comment)

Commit:
Updated value of atomic mass constant to be consistent with 2018 CODATA value.

CameronKing commented:
@asmeurer sorry for missing that, I did get caught up in how to include the uncertainty. I just made the correction you asked for.

bjodah commented:
Regarding the uncertainty, note that one does not just "concatenate" the numbers in parenthesis, they are shorthand for +/-, see e.g. https://physics.nist.gov/cgi-bin/cuu/Info/Constants/definitions.html

And adding the uncertainty to all non-exact constants sounds like a big separate change suitable for a separate PR in either case? (I wonder if the uncertainties are assumed to be normal distributed, I haven't checked)

As I said earlier, I'm not too happy about mixing constants from different revisions: it will "break" in the sense that a user deriving some values using sympy may end up with a different answer than people using a single CODATA revision (albeit the problem is probably academic since few real experiments ever come close to this kind of precision).

asmeurer commented:
Couldn't there also be an inconsistency if constants are related to one another via some of the conversion factors? Or are all the constants independent?

Commit:
Performed a full update of physical constants to 2018 CODATA values

This pull request finishes a full update of the physical constants
to reflect the current 2018 CODATA standard including adding a physical
constant representing the elementary charge. In addition, derived
constants were modified so that they use their underlying definition
rather than their numerical values e.g. von Klintzing constant
is now defined as planck/elementary_charge^2.

bjodah commented:
Some are derived from others, I don't know if any such relations changed in the last revision.

asmeurer commented:
Although if some constants are derived from others it might be better to just write them in the file as a derivation, rather than copying the constant. That way we only write down the fundamental constants, and inconsistencies are impossible.

CameronKing commented:
Fortunately none of the definitions of the derived constants changed. However, their numerical values did change. @bjodah I agree that mixing revisions would be confusing, so I'm working on a full update. The update will also define the derived constants (e.g. Stefan-Boltzmann constant) as their underlying defintion.

CameronKing commented:
@asmeurer @bjodah I just pushed an update to the PR with a full update of the physical constants to the 2018 CODATA values as well as setting derived constants using their underlying definitions.

bjodah commented:
Excellent!

oscargus commented:
Great! Thanks for your efforts!
